### PR TITLE
Implement user account deletion

### PR DESCRIPTION
**Pull request by k4m4**
- Allow users to permanently delete their Vanity account.

### CODE REVIEW

File: backend/workers/helpers.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+const { ingestMetrics, sendEmail } = require('./queues');
+
+const getRepeatableJobsByID = async id => {
+	const ingestMetricsJobs = await ingestMetrics.getJobs(['delayed']);
+	const sendEmailJobs = await sendEmail.getJobs(['delayed']);
```
Comment (dionyziz) [inline backend/workers/helpers.js:5]: Include `wait` in addition to `delayed` here.

File: backend/api/controllers/index.js [hunk 2]
```diff
@@ -132,7 +118,7 @@ const ResubscriptionErrors = {
 
 const resubscribe = async (request, response) => {
 	try {
-		const user = await getUserFromRequest(request);
+		const user = await User.findByRequest(request);
 
 		const jobs = await getRepeatableJobsByID(user.id);
 		const jobsArray = Object.keys(jobs).map(key => jobs[key]);
```
Comment (dionyziz) [inline backend/api/controllers/index.js:124]: `Object.values(jobs)`

File: backend/models/user.js [hunk 3]
```diff
@@ -89,6 +90,18 @@ module.exports = (Sequelize, DataTypes) => {
...
 
+	User.findByRequest = request => {
+		const auth = request.headers.authorization;
+		const { token } = JSON.parse(auth);
+		return User.findByToken(token);
+	};
+
+	User.afterDestroy(async (user, _) => {
+		const jobs = await getRepeatableJobsByID(user.id);
+		const jobsToDelete = Object.keys(jobs).map(key => jobs[key]);
```
Comment (dionyziz) [inline backend/models/user.js:101]: `Object.values(jobs)`

File: backend/models/user.js [hunk 4]
```diff
@@ -89,6 +90,18 @@ module.exports = (Sequelize, DataTypes) => {
...
+		jobsToDelete.filter(job => job).forEach(job => job.remove());
```
Comment (dionyziz) [inline backend/models/user.js:102]: Consider rewriting as `.omitBy(_.isUndefined)` for readability

File: backend/workers/queues.js [hunk 5]
```diff
@@ -35,21 +36,32 @@ const sendSampleEmailWorker = async job => {
 	});
 };
 
+const deleteAccountWorker = async job => {
+	const { User } = require('../models');
+	const { userID } = job.data;
+	const user = await User.findByPk(userID);
+	return user.destroy();
+};
+
```
Comment (dionyziz) [inline backend/workers/queues.js:45]: Consider refactoring these workers and moving each worker to their respective model file. You will need to create a separate email model for this purpose.

File: backend/api/routes/index.js [hunk 6]
```diff
@@ -8,5 +8,7 @@ router.post('/preferences/repos', api.updateRepos);
 router.post('/preferences/metric-types', api.updateMetricTypes);
 router.post('/unsubscribe', api.unsubscribe);
 router.post('/resubscribe', api.resubscribe);
+router.post('/delete-account', api.destroy);
+router.post('/cancel-deletion', api.cancelDestruction);
```
Comment (dionyziz) [inline backend/api/routes/index.js:12]: Use REST

File: backend/workers/helpers.js [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+
+	return {
+		ingestMetrics: await ingestMetricsJobs.find(delayedJob => delayedJob.opts.repeat.jobId === id),
+		sendEmail: await sendEmailJobs.find(delayedJob => delayedJob.opts.repeat.jobId === id),
```
Comment (dionyziz) [inline backend/workers/helpers.js:9]: DRY

File: frontend/components/AccountDeleted.js [hunk 8]
```diff
@@ -0,0 +1,58 @@
+import React, { useState } from 'react';
```
Comment (dionyziz) [inline frontend/components/AccountDeleted.js:1]: Consider renaming your JSX files to have the `.jsx` extension. [See relevant discussion](https://github.com/airbnb/javascript/pull/985)

File: frontend/components/AccountDeleted.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+		return (
+			<Layout>
+				<Main>
+					<h3>your account has been deleted</h3>
+					<p>We are sorry to see you go!</p>
+				</Main>
+				<Footer />
+			</Layout>
+		);
+	}
```
Comment (dionyziz) [inline frontend/components/AccountDeleted.js:31]: To reduce duplication, consider something like this:

```
let classNames = 
let undoButton = <a className='...' onClick={handleCancellation}>undo</a>;

if (!canRecover) {
  undoButton.classNames += ' hidden'
}

return (<Layout><Main> sorry to see you go... {undoButton} </Main></Layout>)
```

File: frontend/pages/preferences.js [hunk 10]
```diff
@@ -1,20 +1,44 @@
...
+		const deleted = await deleteAccount(token);
+		if (deleted) {
+			setAccountDeleted(true);
+		}
+
+		NProgress.done();
+	};
+
+	if (accountDeleted) {
+		COOKIES.forEach(cookie => Cookies.remove(cookie));
```
Comment (dionyziz) [inline frontend/pages/preferences.js:39]: Move to above function so that this line doesn't run repeatedly

File: frontend/logic/preferences.js [hunk 11]
```diff
@@ -73,3 +73,43 @@ export async function resubscribe(token) {
...
+
+export async function cancelAccountDeletion(token) {
+	try {
+		const client = new APIClient(token);
+		const response = await client.post('/api/cancel-deletion');
+
+		if (response.ok) {
+			const { message } = await response.json();
+			await Router.push('/login');
+			toast.success(message, {
```
Comment (dionyziz) [inline frontend/logic/preferences.js:101]: message + '\nPlease login again'

File: frontend/logic/preferences.js [hunk 12]
```diff
@@ -73,3 +73,43 @@ export async function resubscribe(token) {
...
+export async function deleteAccount(token) {
+	try {
+		const client = new APIClient(token);
+		const response = await client.post('/api/delete-account');
+
+		if (response.ok) {
+			return true;
+		}
+
+		return false;
```
Comment (dionyziz) [inline frontend/logic/preferences.js:86]: `return response.ok`

File: frontend/logic/preferences.js [hunk 13]
```diff
@@ -73,3 +73,43 @@ export async function resubscribe(token) {
...
+			const { message } = await response.json();
+			await Router.push('/login');
+			toast.success(message, {
+				className: 'avenir bg-blue center pa3 lh-copy',
+			});
+			return;
+		}
+
+		const { errors } = await response.json();
+		toast.error(errors.message, {
```
Comment (dionyziz) [inline frontend/logic/preferences.js:108]: Rename `errors` to `error` or use an array in your backend.

File: backend/api/controllers/index.js [hunk 14]
```diff
@@ -151,10 +137,60 @@ const resubscribe = async (request, response) => {
...
+	INVALID_TOKEN: DeletionError('User token is invalid'),
+	NONEXISTENT_USER_TO_DELETE: DeletionError('The user that you are trying to delete does not exist'),
+	NONEXISTENT_USER_TO_RECOVER: DeletionError('The user that you are trying to recover does not exist')
+};
+
+const destroy = async (request, response) => {
+	try {
+		const user = await User.findByRequest(request);
+		if (user) {
+			user.userScheduler = new UserScheduler();
```
Comment (dionyziz) [inline backend/api/controllers/index.js:150]: Make this the default scheduler -- if scheduler is undefined, have the user model instantiate the `UserScheduler`.

File: backend/api/controllers/index.js [hunk 15]
```diff
@@ -151,10 +137,60 @@ const resubscribe = async (request, response) => {
...
+			user.userScheduler.scheduleDeletionOfUser(user);
+			return response.status(OK).json({
+				message: `Successfully scheduled deletion of user ${user.username}`
+			});
+		}
+
+		return response.status(NOT_FOUND).json(DeletionErrors.NONEXISTENT_USER_TO_DELETE);
```
Comment (dionyziz) [inline backend/api/controllers/index.js:157]: Errors to top

File: backend/test/api/index.test.js [hunk 16]
```diff
@@ -331,3 +335,89 @@ test.serial('POST /api/resubscribe return error when already subscribed', async
...
+	scheduleDeletionOfUser.returns();
+
+	const response = await request(app)
+		.post('/api/delete-account')
+		.set('authorization', JSON.stringify({ token }));
+
+	t.is(response.status, OK);
+	t.is(response.body.message, `Successfully scheduled deletion of user ${username}`);
+
+	t.true(scheduleDeletionOfUser.calledOnce);
```
Comment (dionyziz) [inline backend/test/api/index.test.js:360]: check it has been called with correct args

File: backend/test/api/index.test.js [hunk 17]
```diff
@@ -331,3 +335,89 @@ test.serial('POST /api/resubscribe return error when already subscribed', async
...
+	t.is(response.status, UNAUTHORIZED);
+	t.is(response.body.errors.message, 'User token is invalid');
+});
+
+test.serial('POST /api/cancel-deletion removes deleteAccount job - with appropriate token', async t => {
+	const { id, email, username } = t.context.user;
+	const user = { id, email };
+	const token = generateToken(user);
+
+	await deleteAccountJob(user);
```
Comment (dionyziz) [inline backend/test/api/index.test.js:387]: Consider renaming to `createAccountDeletionJob`

File: backend/test/workers/helpers.test.js [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+const jobs = require('../../workers/jobs');
+const { getRepeatableJobsByID } = require('../../workers/helpers');
+const { USER } = require('../__fixtures__');
+
+test('getRepeatableJobsByID returns ingestMetrics and sendEmail jobs', async t => {
+	const { id, username, email } = USER;
+	const user = { id, username, email };
+
+	const ingestMetricsJob = await jobs.ingestMetricsJob(user);
+	const sendEmailJob = await jobs.sendEmailJob(user);
```
Comment (dionyziz) [inline backend/test/workers/helpers.test.js:11]: `createIngestMetricsJob`, `createSendEmailJob`

File: backend/test/workers/jobs.test.js [hunk 19]
```diff
@@ -77,3 +79,17 @@ test('sendSampleEmailJob creates job', async t => {
...
 	t.is(taskedJob, undefined);
 });
+
+test('deleteAccountJob creates job', async t => {
+	const job = await deleteAccountJob(user);
+	t.is(job.returnvalue, null);
+	t.is(job.name, '__default__');
+	t.is(job.opts.attempts, QUEUE_ATTEMPTS.DELETE_ACCOUNT);
+	t.deepEqual(job.data, { userID: user.id });
+	t.regex(job.id, /\d+/);
```
Comment (dionyziz) [inline backend/test/workers/jobs.test.js:89]: check eq with `user.id`

File: backend/test/workers/queues.test.js [hunk 20]
```diff
@@ -86,3 +88,23 @@ test('sendSampleEmailWorker calls metrics.fetchCurrent() and email.sendSample()'
 	fetchCurrent.restore();
 	sendSample.restore();
 });
+
+test('deleteAccountWorker calls User.findByPk() and User.prototype.destroy()', async t => {
+	const findByPk = sinon.stub(User, 'findByPk');
+	findByPk.returns(Promise.resolve(new User()));
+
```
Comment (dionyziz) [inline backend/test/workers/queues.test.js:95]: Don't test this

File: backend/api/controllers/index.js (hunk 2 above)
Comment (k4m4) [reply backend/api/controllers/index.js:124]: Unrelated to this PR. Will refactor at a later stage.

File: backend/workers/queues.js (hunk 5 above)
Comment (k4m4) [reply backend/workers/queues.js:45]: Will move worker behaviors into their respective models at a later stage.

File: backend/api/routes/index.js (hunk 6 above)
Comment (k4m4) [reply backend/api/routes/index.js:12]: This refactoring will take place at a later stage and will include all routes (not just those related to this PR).

File: frontend/logic/preferences.js (hunk 13 above)
Comment (k4m4) [reply frontend/logic/preferences.js:108]: Unrelated to this PR. Will address later.

File: backend/test/api/index.test.js (hunk 17 above)
Comment (k4m4) [reply backend/test/api/index.test.js:387]: Will rename all methods that generate jobs as recommended, after this PR has been merged.

File: backend/test/workers/jobs.test.js (hunk 19 above)
Comment (k4m4) [reply backend/test/workers/jobs.test.js:89]: Must modify all job tests, so this correction will happen at a later stage.

File: frontend/components/AccountDeleted.js (hunk 8 above)
Comment (k4m4) [reply frontend/components/AccountDeleted.js:1]: It seems that `.js` is the norm when using Next.js, even when pages and components include `JSX`.

File: backend/test/api/index.test.js (hunk 16 above)
Comment (k4m4) [reply backend/test/api/index.test.js:364]: I don't think this is really necessary here. Plus, I would have to create a new Sequelize User instance to do this.

File: backend/api/controllers/index.js (hunk 14 above)
Comment (k4m4) [reply backend/api/controllers/index.js]: Will implement this improvement at a later stage.

File: frontend/logic/preferences.js (hunk 13 above)
Comment (k4m4) [reply frontend/logic/preferences.js:103]: Fixed in [e655bd5](https://github.com/meeshkan/vanity/commit/e655bd57d43a60778277e033d341a37dfbfb3bf6)

File: backend/api/routes/index.js (hunk 6 above)
Comment (k4m4) [reply backend/api/routes/index.js:12]: Addressed in [f290b16](https://github.com/meeshkan/vanity/commit/f290b16908b97f6819a40db530a9bbef889ee928).

**Pull request merged**